src/coprocessor/circom.rs

```diff
@@ -0,0 +1,164 @@
+    /// for the `circom` binary; in this file you must declare your circom main component.
+    fn print_error(name: &str, available: Vec<String>) -> Result<()> {
+        let available = available.join("\n    ");
+        bail!(
+            "
+{}: no circom gadget named `{name}`.
+Available circom gadgets:
+
+    {available}
+
+If you want to setup a new circom gadget `{name}`, place your circom files in a designated folder and
+create a file called `{name}.circom`. The circom binary expects `{}_FOLDER>/{name}.circom` 
+as the input file; in this file you must declare your circom main component.
+
+Then run `lurk coprocessor --name {name} <{}_FOLDER>` to instansiate a new gadget `{name}`.",
+            Red.bold().paint("error"),
+            name.to_ascii_uppercase(),
+            name.to_ascii_uppercase(),
+        );
+    }
+
+    fn validate_gadget<F: LurkField, C: CircomGadget<F>>(gadget: &C) -> Result<()> {
+        if !circom_dir().exists() {
+            std::fs::create_dir_all(circom_dir())?;
+            return print_error(gadget.name(), vec![]);
+        }
+
+        let name = gadget.name();
+        let circom_folder = circom_dir().join(name);
+
+        if circom_folder.exists() {
+            return Ok(());
+        };
+
+        let mut subdirs = Vec::new();
+
+        for entry in read_dir(circom_dir())? {
+            let entry = entry?;
+            let path = entry.path();
+
+            if path.is_dir() {
+                if let Some(dir_name) = path.file_name() {
+                    if let Some(dir_name) = dir_name.to_str() {
+                        subdirs.push(dir_name.to_string());
+                    }
+                }
+            }
+        }
+
+        if subdirs.contains(&gadget.name().into()) {
+            return Ok(());
+        }
+
+        print_error(gadget.name(), subdirs)
+    }
+
+    #[derive(Debug)]
+    pub struct CircomCoprocessor<F: LurkField, C: CircomGadget<F>> {
+        gadget: C,
+        config: CircomConfig<F>,
+    }
+
+    impl<F: LurkField, C: CircomGadget<F>> Clone for CircomCoprocessor<F, C> {
+        fn clone(&self) -> Self {
+            CircomCoprocessor::new(self.gadget.clone())
+        }
+    }
+
+    impl<F: LurkField, C: CircomGadget<F>> CoCircuit<F> for CircomCoprocessor<F, C> {
+        /// TODO: Generalize
+        fn arity(&self) -> usize {
+            0
+        }
+
+        fn synthesize<CS: ConstraintSystem<F>>(
+            &self,
+            cs: &mut CS,
+            g: &GlobalAllocations<F>,
+            _store: &Store<F>,
+            input_exprs: &[AllocatedPtr<F>],
+            input_env: &AllocatedPtr<F>,
+            input_cont: &AllocatedContPtr<F>,
+        ) -> Result<(AllocatedPtr<F>, AllocatedPtr<F>, AllocatedContPtr<F>), SynthesisError>
+        {
+            let input = self.gadget.clone().into_circom_input(input_exprs);
+            let witness = circom_scotia::calculate_witness(&self.config, input, true).expect("msg");
```

Comment:
I think we can do better than `.expect(“msg”)` specially for a function that returns `Result`